score.go

```diff
@@ -15,31 +15,252 @@
+			if startMatch && endMatch {
+				addScore("Base", scoreBase)
+			} else if startMatch || endMatch {
+				addScore("EdgeBase", (scoreBase+scorePartialBase)/2)
+			} else if sep < int(m.byteOffset) {
+				addScore("InnerBase", scorePartialBase)
+			}
+		} else if sec, si, ok := p.findSymbol(m); ok {
+			startMatch := sec.Start == m.byteOffset
+			endMatch := sec.End == endOffset
+			if startMatch && endMatch {
+				addScore("Symbol", scoreSymbol)
+			} else if startMatch || endMatch {
+				addScore("EdgeSymbol", (scoreSymbol+scorePartialSymbol)/2)
+			} else {
+				addScore("OverlapSymbol", scorePartialSymbol)
+			}
+
+			// Score based on symbol data
+			if si != nil {
+				symbolKind := ctags.ParseSymbolKind(si.Kind)
+				sym := sectionSlice(data, sec)
+
+				addScore(fmt.Sprintf("kind:%s:%s", language, si.Kind), scoreSymbolKind(language, filename, sym, symbolKind))
+
+				// This is from a symbol tree, so we need to store the symbol
+				// information.
+				if m.symbol {
+					if symbolInfo == nil {
+						symbolInfo = make([]*Symbol, len(ms))
+					}
+					// findSymbols does not hydrate in Sym. So we need to store it.
+					si.Sym = string(sym)
+					symbolInfo[i] = si
+				}
+			}
+		}
+
+		// scoreWeight != 1 means it affects score
+		if !epsilonEqualsOne(m.scoreWeight) {
+			score = score * m.scoreWeight
+			if opts.DebugScore {
+				what += fmt.Sprintf("boost:%.2f, ", m.scoreWeight)
+			}
+		}
+
+		if score > bestLine.score {
+			bestLine.score = score
+			bestLine.debugScore = what
+		}
+	}
+
+	if opts.DebugScore {
+		bestLine.debugScore = fmt.Sprintf("score:%.2f <- %s", bestLine.score, strings.TrimSuffix(bestLine.debugScore, ", "))
+	}
+
+	return bestLine, symbolInfo
+}
+
+// scoreLineBM25 computes the score of a line according to BM25, the most common scoring algorithm for text search:
+// https://en.wikipedia.org/wiki/Okapi_BM25. Compared to the standard scoreLine algorithm, this score rewards multiple
+// term matches on a line.
+// Notes:
+// - This BM25 calculation skips inverse document frequency (idf) to keep the implementation simple.
+// - It uses the same calculateTermFrequency method as BM25 file scoring, which boosts filename and symbol matches.
+func (p *contentProvider) scoreLineBM25(ms []*candidateMatch, lineNumber int) (float64, []*Symbol) {
+	// If this is a filename, then don't compute BM25. The score would not be comparable to line scores.
+	if lineNumber < 0 {
+		return 0, nil
+	}
+
+	// Use standard parameter defaults used in Lucene (https://lucene.apache.org/core/10_1_0/core/org/apache/lucene/search/similarities/BM25Similarity.html)
+	k, b := 1.2, 0.75
+
+	// Calculate the length ratio of this line. As a heuristic, we assume an average line length of 100 characters.
+	// Usually the calculation would be based on terms, but using bytes should work fine, as we're just computing a ratio.
+	data := p.data(false)
+	nl := p.newlines()
+	lineLength := len(nl.getLines(data, lineNumber, lineNumber+1))
```

Comment:
I think we'll always have loaded the data by this point so it might not matter, but you can get the line length without the actual data with `nl.lineStart(lineNumber + 1) - nl.lineStart(lineNumber)`